### PR TITLE
added local cache type and mode options

### PR DESCRIPTION
**Pull request by imightybigman**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/7643

Changes proposed in this pull request:

* Add `mode` to `cache` in `aws_codebuild_project`

Output from acceptance testing:

```
 make testacc  TESTARGS='-run=TestAccAWSCodeBuildProject_Cache'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSCodeBuildProject_Cache -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSCodeBuildProject_Cache
=== PAUSE TestAccAWSCodeBuildProject_Cache
=== CONT  TestAccAWSCodeBuildProject_Cache
--- PASS: TestAccAWSCodeBuildProject_Cache (68.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	68.265s
```

### CODE REVIEW

**Comment by imightybigman** [comment]
First time making an addition to Terraform if there are any obvious improvements, please let me know before I attempt the acceptance tests

**Comment by imightybigman** [comment]
Test updated

**Comment by AlmogCohen** [comment]
That's such an important feature on AWS CodeBuild to make the build way quicker for docker...eager to see it integrated

**Comment by jbienkowski311** [comment]
@bflad any chance on getting it merged soon? It's a pretty cool feature

**Comment by jbienkowski311** [comment]
@imightybigman the documentation part is missing in your PR. Please check the `website` directory in the repository and add missing value for CodeBuild cache type

**Comment by imightybigman** [comment]
ah ok, no problem. I will do that tonight

**Comment by imightybigman** [comment]
@jbienkowski311  the markdown has been updated

**Comment by smiller171** [comment]
Is there anything keeping this from being merged? Would really love to be able to enable docker caching in my Terraform.

**Comment by katieschilling** [comment]
Anything preventing a merge here? We'd love to have this feature. @azalenski-callpage

**Comment by amilstead** [comment]
I'll keep the heartbeat going. @azalenski-callpage is there anything left to do here?

**Comment by mikesigs** [comment]
Would also love to see this merged!

File: aws/resource_aws_codebuild_project.go [hunk 1]
```diff
@@ -99,12 +99,18 @@ func resourceAwsCodeBuildProject() *schema.Resource {
...
 								codebuild.CacheTypeS3,
+								codebuild.CacheTypeLocal,
 							}, false),
 						},
 						"location": {
 							Type:     schema.TypeString,
 							Optional: true,
 						},
+						"modes": {
+							Type:     schema.TypeList,
```
Comment (bflad) [inline aws/resource_aws_codebuild_project.go]: If the API ordering of this attribute is not guaranteed when multiple modes are configured, we will likely need to change this to a `TypeSet`.

File: aws/resource_aws_codebuild_project.go [hunk 2]
```diff
@@ -99,12 +99,18 @@ func resourceAwsCodeBuildProject() *schema.Resource {
...
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
```
Comment (bflad) [inline aws/resource_aws_codebuild_project.go]: One future enhancement that can be done here is supporting validation using the available [AWS Go SDK constants](https://docs.aws.amazon.com/sdk-for-go/api/service/codebuild/#pkg-constants):

```suggestion
							Elem:     &schema.Schema{
								Type: schema.TypeString,
								ValidateFunc: validation.StringInSlice([]string{
									codebuild.CacheModeLocalDockerLayerCache,
									codebuild.CacheModeLocalSourceCache,
									codebuild.CacheModeLocalCustomCache,
								},false),
							},
```

**Comment by bflad** [comment]
This has been released in [version 2.12.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2120-may-24-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**